Presburger set review

Dinistro opened the pull request:
It's not intended to merge this yet. But we need a place to discuss the PresburgerSet review. 

Note: I think there is currently a bug in complement, I will look into this. The rest should be more or less fine for now.

Dinistro committed:
feature: Presburger dialect prototype

Dinistro committed:
feature: first attempt on a CI

Dinistro committed:
fix: CI now only builds and tests + removal of presburgerc

Dinistro committed:
feature: PresburgerSet subtraction prototype

This contains the first prototype for the Presburger set subtraction. It
is taken from libint but does not support the redundant optimization.

Dinistro committed:
fix: recrusive subtraction now actually recurses

After discussion with Arjun it turned out that the libint implementation
was faulty, therefore we had to change the one here.

Dinistro committed:
feature: subtraction operation prototype

Dinistro committed:
fix: reverted reformated comments

It turned out that clang-format was executed with the wrong config once.

Dinistro committed:
fix: CI change cache id

Dinistro committed:
fix: PresburgerSet subtraction now supports equalies properly

Dinistro committed:
testing: canonicalization tests for set subtraction

Dinistro committed:
chore: small comments for PresburgerSet

Dinistro committed:
feature: PresburgerSet complement

Dinistro committed:
feature: presburger.complement operation + canonicalization

Dinistro committed:
fix: PresburgerSet subtraction added missing inequality to FlatAffineConstraint

Dinistro commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,349 @@
+  if (markedEmpty) {
+    // TODO dicuss what we want to print in the empty case
+    // os << ": (1 = 0)";
+    // return;
```
@Superty In libint you just add `"<marked empty>"` do you suggest that I do the same here? Or is a trivially false constraint better as it can be parsed without additional work. 
Note that this is probably broken in libint as I was not able to find anything with "empty" in the parser.

superty replied:
I think a trivially false constraint is fine.

Yes, the libint parser can't parse this at the moment.

Dinistro committed:
fix: PresburgerSet::subtract now again uses a ref

Dinistro committed:
fix: print (1 = 0) for empty set

If a set is marked empty, we better print it as empty. This was a
problem for sets without constraints that were marked empty.

Superty commented:
It would be nice to have more documentation in `Types.h` and `Set.h`. What is the difference between dimensions and symbols, and do we need to distinguish between the two in `PresburgerSet`? For the purpose of analysis and operations on sets, there does not appear to be any distinction between them.

Dinistro commented:
Agreed, I will try to add more documentation. 
I based the design on FlatAffineConstraints that also supports symbols. Currently we cannot set symbols, but they would allow us to make a Set dependent on a SSA value (like for example a memref size). This is partly described in https://mlir.llvm.org/docs/Dialects/Affine/

Dinistro committed:
refactoring: add constraint transformation funcs

Dinistro committed:
feature: PresburgerSet integer sampling

Dinistro commented:
I think symbols need more discussion, see https://github.com/Superty/llvm-project/issues/8

Dinistro committed:
feature: check that no symbols are present

As we do currently not support symbols in all cases, we have to assert
that non are present before we try to perform operations.

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+  assert(cs.getNumDimIds() == nDim &&
+         "Cannot add FlatAffineConstraints having different dimensionality");
+
+  if (cs.isEmptyByGCDTest())
```
Hm, not sure if we can say a prioiri that this heuristic helps more than it costs. It might be better to leave it out for now.

Also, the number of symbols needs to be equal to that of the `Set`. I would suggest making something like a  `checkDimensionsCompatible` that checks both, as it will be used in almost every operation.

Dinistro replied:
I have no experience with these heuristics, so I will do so. 

Adding a `checkDimensionsCompatible` makes sense to me

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+void PresburgerSet::unionSet(const PresburgerSet &set) {
+  assert(set.getNumDims() == nDim &&
+         "Cannot union Presburger sets having different dimensionality");
+  assert(set.getNumSyms() + nSym == 0 &&
```
This implementation of union already works for parameters. Also below for intersection.

superty replied:
Also, check for equality of number of symbols.

Dinistro replied:
This works if we ignore the naming of the parameters/symbols. If we are only considering the position of the symbol we can do this nicely, otherwise we need more functionality to do so.

Dinistro replied:
I'm doing it as you suggested for now.

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+void PresburgerSet::intersectSet(const PresburgerSet &set) {
+  assert(set.getNumDims() == nDim && set.getNumSyms() == nSym &&
+         "Cannot intersect Presburger sets having different dimensionality");
+  assert(set.getNumSyms() + nSym == 0 &&
```
This implementation of intersection already works.

Dinistro replied:
Same as above.

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+
+SmallVector<int64_t, 64> inequalityFromEquality(const ArrayRef<int64_t> &eq,
+                                                bool negated, bool strict) {
+  SmallVector<int64_t, 64> coeffs;
```
The average dimensionality doesn't seem to go beyond 25 from what we've seen, so we can probably decrease the length parameter here.

Dinistro replied:
I will do so, good point.

Dinistro replied:
What number do you suggest? 25? Or just something like 16?

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+
+  // The constant is at the end
+  if (strict)
+    --coeffs[eq.size() - 1];
```
`--coeffs.back()` might be cleaner?

Dinistro replied:
Agreed

superty commented on mlir/lib/Analysis/Presburger/Simplex.cpp:
```diff
@@ -567,6 +567,15 @@ llvm::Optional<std::vector<int64_t>> Simplex::getSamplePointIfIntegral() const {
 }
 
+void Simplex::addFlatAffineConstraints(const FlatAffineConstraints &cs) {
+  assert(cs.getNumDimIds() + cs.getNumSymbolIds() == numberVariables() &&
```
I believe this should be `cs.getNumIds() == numberVariables()`. (the local variables are also be included)

Dinistro replied:
Good point

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+  if (i == S.getNumBasicSets()) {
+    FlatAffineConstraints BCopy = B;
+    // BCopy.simplify();
+    result.addFlatAffineConstraints(std::move(BCopy));
```
No need to copy if we're not simplifying now.

Dinistro replied:
I forgot that the constraints will be copied, you are right, thanks

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+    simplex.rollback(snap);
+  };
+
+  size_t addedIneqs = 0;
```
Might be cleaner to store the number of inequalities present at the start and remove all the ones added beyond that later, instead of keeping track of how many were added.

Dinistro replied:
Correct. Furthermore it should be of type `unsigned`, not `size_t`.

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+    // TODO reimplement the heuristics
+    const auto &eq = S_i.getEquality(j);
+    recurseWithInequalityFromEquality(eq, true, true);
+    recurseWithInequalityFromEquality(eq, false, true);
```
Looking at this now, I feel it would make more sense to have a `recurseWithInequality(ineq)` that is called from here with the inequality `inequalityFromEquality(eq, ...)`. We can then reuse the same function below with the inequality `complement`. 

Instead of the two `addInequalityFromEquality(...)` calls below we just need to add the equality itself to `B` and `simplex.` (and remove the equalities from `B` as well at the end)

Dinistro replied:
I already rewrote this partially in the `feature-equality` branch as we have the possibility to use `isMarkedEmpty` for the heuristic. I will add the remaining things.

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+}
+
+// Returns the set difference B - S.
+PresburgerSet PresburgerSet::subtract(FlatAffineConstraints c,
```
The function argument is called `c` but is referred to as `B` in the documentation.

Dinistro replied:
Good catch, I will change this

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+  if (set.isMarkedEmpty())
+    return PresburgerSet(set.getNumDims(), set.getNumSyms());
+
+  PresburgerSet result(c.getNumDimIds());
```
Need to pass the number of symbols here.

Dinistro replied:
Yes. With the other changes it's now possible to reach this with `set.getNumSyms() != 0`

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+  if (set.isUniverse())
+    return PresburgerSet::makeEmptySet(set.getNumDims(), set.getNumSyms());
+  if (set.isMarkedEmpty())
+    return PresburgerSet(set.getNumDims(), set.getNumSyms());
```
If `set` is empty, the result should be `c`, right?

Dinistro replied:
This is correct. The tests do not detect this, as this check can only be reached if c is universe. I will fix it though

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+    return;
+  }
+  if (isUniverse()) {
+    *this = PresburgerSet::complement(set);
```
This is unnecessary, right? The same thing will happen below.

Dinistro replied:
I don't think so. If `isUniverse()` is true, then `flatAffineConstraints` might be empty. In this case we will just return `result` which is an empty set.

superty replied:
Yes, you're right.

superty commented on mlir/lib/Dialect/Presburger/Set.cpp:
```diff
@@ -0,0 +1,396 @@
+llvm::hash_code PresburgerSet::hash_value() const {
+  // TODO how should we hash FlatAffineConstraints without having access to
+  // private vars?
+  return llvm::hash_combine(nDim, nSym);
```
It's enough if `FlatAffineConstraints` provides a `hash_value`, right? We can then `hash_combine` the list of FACs. This should be fine once we switch to our `PresburgerBasicSet`; we can provide this functionality there.

Dinistro replied:
That's correct, I added this already as the PresburgerSetAttribute needs the hash of the PresburgerSet and the FlatAffineConstraints do not support hashing yet.

superty reviewed (commented):
I left some comments on `PresburgerSet` and the changes to `Simplex`. 

I now feel that it might be cleaner to instead switch to viewing a Set as a union of FACs, so that an empty FAC list would be interpreted as an empty Set. This may be much more natural to work with. We can represent `Universe` with a single trivial constraint, `0 = 0`. The only question would be how this should be printed, because an empty list of constraints looks like a Universe rather than an empty set. We can actually parse something like `<empty>` or we can use a canonical impossible constraint like `1 == 0` as you mentioned.

Also, the variable names don't follow LLVM conventions. `clang-tidy` would probably point out all such issues.

Dinistro commented:
All the suggestions were implemented in the feature-equality branch. If there is nothing more, I will close this review PR.

Superty commented:
Sounds good.
